kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala

```diff
@@ -484,6 +485,182 @@ sealed abstract class Resource[F[_], +A] {
+              else
+                s
+            }
+          } flatMap {
+            // if the inner F has a zero, we lose the finalizers, but there's no avoiding that
+            case (a, rel) =>
+              val action = state modify { s =>
+                if (s.confirmedFinalizeOnComplete)
+                  (s, rel.handleError(_ => ()))
+                else
+                  (s.copy(fin = rel), F.unit)
+              }
+
+              action.flatten.as(a)
+          }
+        }
+
+        F.start(finalized) map { outer =>
+          val fiber = new Fiber[Resource[F, *], Throwable, A] {
+            def cancel =
+              Resource eval {
+                F uncancelable { poll =>
+                  // technically cancel is uncancelable, but separation of concerns and what not
+                  poll(outer.cancel) *> state.update(_.copy(finalizeOnComplete = true))
+                }
+              }
+
+            def join =
+              Resource eval {
+                outer.join.flatMap[Outcome[Resource[F, *], Throwable, A]] {
+                  case Canceled() =>
+                    Outcome.canceled[Resource[F, *], Throwable, A].pure[F]
+
+                  case Errored(e) =>
+                    Outcome.errored[Resource[F, *], Throwable, A](e).pure[F]
+
+                  case Succeeded(fp) =>
+                    state.get map { s =>
+                      if (s.confirmedFinalizeOnComplete)
+                        Outcome.canceled[Resource[F, *], Throwable, A]
+                      else
+                        Outcome.succeeded(Resource.eval(fp))
+                    }
+                }
+              }
+          }
+
+          val finalizeOuter =
+            state.modify(s => (s.copy(finalizeOnComplete = true), s.fin)).flatten
+
+          (fiber, finalizeOuter)
+        }
+      }
+    }
+  }
+
+  def evalOn(ec: ExecutionContext)(implicit F: Async[F]): Resource[F, A @uncheckedVariance] =
+    Resource applyFull { poll =>
+      poll(this.allocated).evalOn(ec) map { p =>
+        Functor[(A @uncheckedVariance, *)].map(p)(fin => (_: Resource.ExitCase) => fin)
+      }
+    }
+
+  def attempt[E](implicit F: MonadError[F, E]): Resource[F, Either[E, A]] =
+    this match {
+      case Allocate(resource) =>
+        Resource.applyFull { poll =>
+          resource(poll).attempt.map {
+            case Left(error) => (Left(error), (_: ExitCase) => F.unit)
+            case Right((a, release)) => (Right(a), release)
+          }
+        }
+      case Bind(source, f) =>
+        MonadError[Resource[F, *], E].attempt(source).flatMap {
+          case Left(error) => Resource.pure(error.asLeft)
+          case Right(s) => f(s).attempt
+        }
+      case p @ Pure(_) =>
+        Resource.pure(p.a.asRight)
+      case e @ Eval(_) =>
+        Resource.eval(e.fa.attempt)
+    }
+
+  def handleErrorWith[E](f: E => Resource[F, A @uncheckedVariance])(
+      implicit F: MonadError[F, E]): Resource[F, A] =
```

Comment:
Actually, this is the compiler telling you that you need a different type here! In particular, the signature should be something like this:

```suggestion
  def handleErrorWith[B >: A, E](f: E => Resource[F, B])(implicit F: MonadError[F, E]): Resource[F, B] =
```

Reply:
Yeah, the compiler was pointing that the type variance wasn't ok.
I've switched to this approach (thank you for pointing it out) and removed the `@uncheckedVariance` occurrences